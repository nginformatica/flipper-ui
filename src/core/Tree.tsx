import React, { Component } from 'react'
import Node from './Node'

export interface INode {
    id: string | number
    name: string
    nodes?: INode[]
}

interface TreeProps {
    nodes?: INode[]
}

class Tree extends Component<TreeProps, {}> {
    public renderNode(node: INode, index: string, root = false) {
        const { id, name, nodes } = node

        return (
            <Node
                id={index}
                name={name}
                key={id || index}
                style={root ? { padding: 0 } : {}}>
                {nodes && nodes.map(this.renderNode.bind(this))}
            </Node>
        )
    }

    public render() {
        return (this.props.nodes || []).map((node, index) =>
            this.renderNode(node, index.toString(), true)
        )
    }
}

export default Tree
