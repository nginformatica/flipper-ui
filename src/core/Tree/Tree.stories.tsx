import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import Tree from '.'

export default {
    title: 'Tree',
    component: Tree
} as ComponentMeta<typeof Tree>

const Template: ComponentStory<typeof Tree> = args => <Tree {...args} />

export const Default = Template.bind({})
Default.args = {
    nodes: [
        {
            id: 0,
            name: 'Tree',
            nodes: [
                {
                    id: 1,
                    name: 'Branch'
                },
                {
                    id: 2,
                    name: 'Branch'
                },
                {
                    id: 3,
                    name: 'Branch',
                    nodes: [
                        {
                            id: 5,
                            name: 'Leaf'
                        }
                    ]
                }
            ]
        }
    ]
}
