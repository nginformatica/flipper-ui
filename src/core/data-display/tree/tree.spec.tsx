import React from 'react'
import { act } from 'react-dom/test-utils'
import { render, screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import Tree from '.'

describe('Tree', () => {
    const NODES = [
        {
            id: 0,
            name: 'Tree',
            nodes: [
                {
                    id: 1,
                    name: 'Branch'
                },
                {
                    id: 2,
                    name: 'Branch'
                },
                {
                    id: 3,
                    name: 'Branch',
                    nodes: [
                        {
                            id: 5,
                            name: 'Leaf'
                        }
                    ]
                }
            ]
        }
    ]

    it('should render the Tree', () => {
        render(<Tree nodes={NODES} />)

        const root = screen.getByText('root')

        expect(root).toBeDefined()
    })

    it('should render with no nodes', () => {
        render(<Tree />)

        const root = screen.getByText('root')

        expect(root).toBeDefined()
    })

    it('should expand one level', async () => {
        render(<Tree nodes={NODES} />)

        const root = screen.getByText('root')

        await act(async () => await userEvent.click(root))

        const tree = screen.getByText('Tree')

        expect(tree).toBeDefined()
    })

    it('should expand two levels', async () => {
        render(<Tree nodes={NODES} />)

        const root = screen.getByText('root')

        await act(async () => await userEvent.click(root))

        const tree = screen.getByText('Tree')

        await act(async () => await userEvent.click(tree))

        const leaf = screen.getAllByText('Branch')

        expect(leaf.length).toEqual(3)
    })

    it('should match snapshot', () => {
        const { container } = render(<Tree nodes={NODES} />)

        expect(container).toMatchSnapshot()
    })
})
