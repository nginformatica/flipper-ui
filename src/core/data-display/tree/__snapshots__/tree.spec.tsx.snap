// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree should match snapshot 1`] = `
<div>
  <ul
    class="sc-aYaIB exChqK"
    name="root"
  >
    <li
      class="sc-gEvDqW hovEjp"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-fqkwJk iYAchi css-i4bv87-MuiSvgIcon-root"
        data-testid="KeyboardArrowDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
        />
      </svg>
      root
    </li>
  </ul>
</div>
`;
